### ISSUE TITLE
Creating CanvasAnimatedControl in code sometimes hangs

### ISSUE THREAD

**Issue by puzzler-1** [Microsoft/Win2D#319]
I'm working on a UWP using Visual Studio Community 2015. I've added Win2D.uwp version 1.16.0. I will use the CanvasAnimatedControl provided by Win2D to display animated emoticons in a RichTextBlock, so I'm wrapping it in a user control, which I'll use for my animation logic.
The problem I'm facing is some random hanging of the application when adding the user control to the RichTextBlock (sometimes on the first insert, sometimes up to the 8th). It doesn't matter whether I use the Mobile Emulator or Local Machine when testing. The app just stops responding, keeping the pressed state of the button. The Mobile Emulator's "hardware" keys do nothing either.

Here is what I log:

```
46_27_754 => Created (0)
46_27_754 => Loaded (0)
46_27_754 => Loaded (0)
46_27_754 => OnCreateResources (0)
46_27_925 => Created (1)
46_27_925 => Loaded (1)
46_27_925 => Loaded (1) // and here the app hangs
```

And the code with wich to reproduce the issue:
A user control:

```
<UserControl ...
    x:Class="RichTextInlineCanvasTest.CanvasControl"
    xmlns:canvas="using:Microsoft.Graphics.Canvas.UI.Xaml">

    <canvas:CanvasAnimatedControl x:Name="canvas" CreateResources="OnCreateResources" ClearColor="#330000ff" />

</UserControl>
```

With code behind:

```
public sealed partial class CanvasControl : UserControl
{
    static uint globalCounter = 0;
    uint _number = 0;

    public CanvasControl()
    {
        this.InitializeComponent();

        Name = DateTime.Now.ToString("mm_ss_fff");
        _number = globalCounter++;

        this.Loaded += OnLoaded;
        this.Unloaded += OnUnloaded;
        Debug.WriteLine(string.Format("{0} => Created ({1})", Name, _number));
    }

    private void OnLoaded(object sender, RoutedEventArgs e)
    {
        Debug.WriteLine(string.Format("{0} => Loaded ({1})", Name, _number));
    }

    private void OnUnloaded(object sender, RoutedEventArgs e)
    {
        Debug.WriteLine(string.Format("{0} => Unloaded ({1})", Name, _number));
    }

    void OnCreateResources(CanvasAnimatedControl sender, CanvasCreateResourcesEventArgs args)
    {
        Debug.WriteLine(string.Format("{0} => OnCreateResources ({1})", Name, _number));
    }
}
```

MainPage is:

```
<StackPanel Background="White">
    <Button Click="Button_Click" Content="Insert canvas"/>
    <RichTextBlock Name="richText"/>
</StackPanel>
```

And code behind:

```
private void Button_Click(object sender, RoutedEventArgs e)
{
    CanvasControl canvas = new CanvasControl();
    canvas.Width = 32;
    canvas.Height = 32;
    InlineUIContainer imageContainer = new InlineUIContainer();
    imageContainer.Child = canvas;
    Paragraph paragraph = new Paragraph();
    paragraph.Inlines.Add(imageContainer);
    richText.Blocks.Add(paragraph);
}
```

**Comment by damyanp** [comment]
CanvasAnimatedControl is built on top of [SwapChainPanel](https://msdn.microsoft.com/en-us/library/windows/apps/windows.ui.xaml.controls.swapchainpanel.aspx).  The general advice here is that you want to have more than 4 on the screen at a time.  Due to the additional game-thread policy that CanvasAnimatedControl adds I wouldn't recommend adding a large number at a time.  For a scenario such as this, I'd suggest using CanvasControl instead.

However, I would still not expect this to result in a hang, so we will see if we can figure out what's going on.

**Comment by puzzler-1** [comment]
Thank you for the advice.
If there is any other information I can provide regarding the issue, let me know.

**Comment by shawnhar** [comment]
I can repro this issue - thanks for the report!

The root cause is that something in InlineUIContainer or Paragraph is calling Loaded and Unloaded recursively, in some cases loading the control multiple times and then unloading it once while expecting it to remain operational - but CanvasAnimatedControl does not handle this gracefully.

We'll fix that in Win2D, but also I second Damyan's point that CanvasAnimatedControl isn't really suitable for this purpose - it's a heavyweight control that you should only ever have a small number of.  For smaller elements within a UI, CanvasControl is a better choice (and CanvasControl works fine inside a Paragraph).

**Comment by puzzler-1** [comment]
I heeded his advice. CanvasControl works nicely in my case. I'll use a single CanvasAnimatedControl only for the emoticon selection, not in a RichTextBlock, so at the end the bug won't affect me.
Thanks and happy coding.

**Comment by shawnhar** [comment]
Glad to hear you are unblocked.  In the meantime, I understand why this was hanging in Win2D and am working on a fix for that.

**Comment by shawnhar** [comment]
This is fixed in Win2D 1.17.0
